src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala

```diff
@@ -69,19 +69,23 @@ object PV63 {
             RLPList(RLPValue(hpEncode(keyNibbles.toArray[Byte], isLeaf = false)), child.fold(_.hash, _.value): RLPEncodeable)
           case n: MptBranch =>
             import n._
-            RLPList(children.map(e => e.fold(_.hash, _.value)).map(e => byteStringEncDec.encode(e)) :+ (terminator: RLPEncodeable): _*)
+            RLPList(children.map(e => e.fold(a => byteStringEncDec.encode(a.hash), b => rawDecode(b.value.toArray[Byte]))) :+ (terminator: RLPEncodeable): _*)
         }
       }
 
       override def decode(rlp: RLPEncodeable): MptNode = rlp match {
         case rlpList: RLPList if rlpList.items.length == BranchNodeChildLength + 1 =>
-          MptBranch(rlpList.items.take(BranchNodeChildLength).map { bytes =>
-            val v = rlpDecode[ByteString](bytes)
-            if (v.nonEmpty && v.length < 32) {
-              Right(MptValue(v))
-            } else {
-              Left(MptHash(v))
-            }
+          MptBranch(rlpList.items.take(BranchNodeChildLength).map {
+            case bytes: RLPValue =>
+              val v = rlpDecode[ByteString](bytes)
+              if (v.nonEmpty && v.length < 32) {
+                Right(MptValue(ByteString(rlpEncode(bytes))))
```

Comment:
IMO it would be simpler to have only hashes as children of node, I do not think that this optimisation has big impact on tree size, but it deffinitly impacts complexity